File: research/results/native_cuda_2026_09_10/audit.py
"""Recheck the supplied 06a4379 report bundle; no model execution or downloads.

Validates report integrity and aggregate guards, not KL recomputation from
missing probe tensors. Run with the extracted original-reports.tar.gz folder.
"""
from __future__ import annotations

import argparse
import hashlib
import json
import math
import re
import subprocess
from pathlib import Path, PurePosixPath

REPO = Path(__file__).resolve().parents[3]
REVISION = "06a4379c71074e4606116c835941388d1876599b"


def identity(path):
    data = path.read_bytes()
    return {"bytes": len(data), "sha256": hashlib.sha256(data).hexdigest()}


def producer_file(name):
    # Reproduce the historical audit after workflow code has evolved. Compare
    # the recorded producer revision, not today's edited working tree.
    return subprocess.check_output(["git", "show", f"{REVISION}:{name}"], cwd=REPO)


def audit(root):
    def read(name):
        return json.loads((root / name).read_text())
    def report(stage, name="report.json"):
        return read(f"stages/{stage}/attempt-001/{name}")
    files = {}
    for p in sorted(root.rglob("*")):
        if p.is_symlink():
            raise ValueError(f"Unexpected symlink: {p.name}")
        if p.is_file():
            files[p.relative_to(root).as_posix()] = identity(p)
    workflow, summary = read("workflow.json"), read("summary.json")
    controls = workflow["controls"]
    assert controls["revision"] == REVISION and controls["dirty_local_test"] is False
    assert controls["backend"] == "CUDA" and controls["arms"] == ["b5_v6_s0"]
    assert controls["synthetic_only"] is False and controls["timing"] is False
    assert len(workflow["attempts"]) == len(summary["stages"]) == 11
    assert workflow["status"] == summary["status"] == "passed" and summary["error"] is None
    for attempt, row in zip(workflow["attempts"], summary["stages"], strict=True):
        assert attempt["name"] == row["name"] and attempt["status"] == row["status"] == "passed"
        assert attempt["elapsed_seconds"] == row["elapsed_seconds"]
    total = sum(s["elapsed_seconds"] for s in summary["stages"])
    assert math.isclose(total, workflow["active_seconds"], abs_tol=1e-8)
    assert math.isclose(total / 60, summary["active_minutes"], abs_tol=1e-8)
    producer_root = PurePosixPath(workflow["attempts"][0]["directory"]).parents[2]
    checked, omitted = [], set()
    for attempt in workflow["attempts"]:
        for name, expected in attempt["artifacts"].items():
            path = PurePosixPath(name)
            if path.is_relative_to(producer_root):
                relative = path.relative_to(producer_root).as_posix()
                assert files[relative] == expected, relative
                checked.append(relative)
            else:
                omitted.add(name)

    build = report("native-build-and-load", "build-receipt.json")
    assert build["root_revision"] == REVISION and build["root_dirty"] is False
    assert build["backend"] == "CUDA" and build["load_validated"] is True
    code = {**controls["workflow_sources"], **build["external_sources"]}
    for name, expected in code.items():
        assert hashlib.sha256(producer_file(name)).hexdigest() == expected, name
    contract = json.loads(producer_file("integrations/llama.cpp/rotquant-native-v2-files.json"))
    assert build["patch_sha256"] == contract["patch_sha256"]
    assert build["base_revision"] == contract["base_revision"]
    build_artifacts = workflow["attempts"][2]["artifacts"]
    for name, expected in contract["files_sha256"].items():
        assert build_artifacts[controls["work_dir"] + "/llama-source/" + name]["sha256"] == expected
    runtime = report("binding-load", "load.json")["runtime_files"]
    assert runtime["librotquant_ggml_test.so"] == build["library_sha256"]
    for name, expected in runtime.items():
        # Workflow resolves symlinks; numerical receipts also name .so aliases.
        family = name.split(".so", 1)[0] + ".so"
        matches = [v for k, v in build_artifacts.items()
                   if PurePosixPath(k).name == family or PurePosixPath(k).name.startswith(family + ".")]
        assert matches and all(v["sha256"] == expected for v in matches), name
    env = report("environment", "environment.json")
    for name in ("bootstrap-before.json", "bootstrap-after.json"):
        bootstrap = report("environment", name)
        assert bootstrap["passed"] and bootstrap["torch"] == env["torch"]
        assert bootstrap["torch_file"] == env["torch_file"]
    operators = {}
    for backend in ("CPU", "CUDA0"):
        data = report("operators-" + backend)
        assert data["backend"] == backend and data["runtime_files"] == runtime
        expected_cases = {(bits, scale, width, tokens) for bits, scale in ((5, 8), (6, 16), (8, 16))
                          for width in (256, 512) for tokens in (1, 7, 33)}
        assert {(c["bits"], c["scale_bits"], c["width"], c["tokens"]) for c in data["cases"]} == expected_cases
        assert len(data["cases"]) == 18 and data["passed"] is True
        assert all(c["passed"] and 0 <= c["max_abs"] <= .001 for c in data["cases"])
        operators[backend] = {"cases": 18, "max_abs": max(c["max_abs"] for c in data["cases"])}
    def guards(data, metrics):
        for key in ("max_abs_error", "mean_abs_error", "mean_kl"):
            assert 0 <= metrics[key] <= data["thresholds"][key]
        minimum = data["thresholds"].get("top1_agreement_min", data["thresholds"].get("top1_agreement"))
        assert minimum <= metrics["top1_agreement"] <= 1
    synthetic = {}
    for bits in (6, 8):
        data = report("whole-model-w" + str(bits))
        assert data["runtime_files"] == runtime and data["backend"] == "CUDA0"
        assert data["passed"] and data["cpu_fallback_forbidden"] and data["gpu_custom_ops"] > 0
        assert all(data["guards"].values()) and set(data["by_prompt_length"]) == {"1", "4", "17", "64"}
        guards(data, data["metrics"])
        for metrics in data["by_prompt_length"].values():
            guards(data, metrics)
        synthetic[f"w{bits}"] = data["metrics"]
    conversion = report("hf-conversion")
    assert conversion["passed"] and conversion["runtime_files"] == runtime
    assert {(c["backend"], c["tokens"]) for c in conversion["cases"]} == {
        (backend, tokens) for backend in ("CPU", "CUDA0") for tokens in (4, 17, 64)}
    for case in conversion["cases"]:
        assert case["passed"]
        guards(conversion, case["metrics"])
    retained = report("retained-b5_v6_s0", "retained/report.json")
    source = report("source-b5_v6_s0", "source.json")
    exported = report("export-b5_v6_s0", "export.json")
    assert retained["passed"] and retained["runtime_files"] == runtime
    assert retained["backend"] == "CUDA0" and retained["cpu_fallback_forbidden"] and retained["gpu_custom_ops"] > 0
    assert retained["checkpoint_manifest_sha256"] == source["checkpoint_sha256"] == exported["checkpoint_manifest_sha256"]
    assert retained["probe_sha256"] == source["probe_sha256"]
    assert retained["export_sha256"] == files["stages/export-b5_v6_s0/attempt-001/export.json"]["sha256"]
    assert retained["export_files"] == exported["artifact_files"]
    size = sum(f["bytes"] for f in exported["artifact_files"].values())
    assert size == retained["complete_model_payload_bytes"] == exported["complete_model_payload_bytes"]
    assert len(exported["matrices"]) == 201
    assert sum(m["vocabulary"] for m in exported["matrices"]) == 1
    for matrix in exported["matrices"]:
        assert (matrix["bits"], matrix["scale_bits"]) == ((6, 16) if matrix["vocabulary"] else (5, 8))
    parity = retained["parity"]
    assert parity["thresholds"] == {"max_abs_error": .125, "mean_abs_error": .005,
                                    "mean_kl": .0001, "top1_agreement_min": .99}
    guards(parity, parity)
    assert all(parity["guards"].values()) and parity["positions"] == 16
    assert parity["generation_probes"] == parity["exact_generation_probes"] == source["probes"] == 4
    log = (root / "stages/retained-b5_v6_s0/attempt-001/logs/run_rq3_retained_gpu.log").read_text()
    probes = re.findall(r"native parity (p\d+): (\d+) input tokens, (\d+) logit positions, (\d+) generated tokens", log)
    assert probes == [(f"p{i}", "64", "4", "8") for i in range(4)]
    assert "offloaded 33/33 layers to GPU" in log and re.search(r"n_ctx\s+= 256\b", log)
    assert retained["timing"] == {"skipped": "not requested"}
    return {"protocol": "native-cuda-report-audit-v1", "reviewed_revision": REVISION,
            "passed_with_caveats": True, "stage_count": 11, "bundle_file_count": len(files),
            "bundle_index": files, "available_receipt_hashes_verified": len(checked),
            "source_code_hashes_verified": len(code), "recorded_patched_source_hashes_matched": len(contract["files_sha256"]),
            "recorded_external_artifacts_not_in_report_bundle": len(omitted),
            "active_minutes": total / 60, "build_minutes": summary["stages"][2]["elapsed_seconds"] / 60,
            "non_build_minutes": (total - summary["stages"][2]["elapsed_seconds"]) / 60,
            "environment": env, "operators": operators, "synthetic_models": synthetic,
            "retained_parity": parity, "model_payload_bytes": size,
            "model_payload_gb": size / 1e9, "model_payload_gib": size / 2**30,
            "text_gguf_bytes": exported["artifact_files"]["model.gguf"]["bytes"],
            "auxiliary_bytes": exported["artifact_files"]["auxiliary.safetensors"]["bytes"],
            "sampled_process_vram_mib": retained["memory"]["sampled_peak_process_vram_mib"],
            "sampled_process_vram_gib": retained["memory"]["sampled_peak_process_vram_mib"] / 1024,
            "vram_samples": retained["memory"]["samples"], "reserved_context": 256,
            "probe_shapes": probes, "timing_requested": False,
            "limits": ["Reported aggregates and hashes checked; absent logits/trace tensors prevent independent numerical recomputation.",
                       "Weights, exported GGUF and native binaries are absent from the reports-only bundle; remote hashes are corroborated, not recomputed locally.",
                       "Four short text prompts, not task accuracy, full-precision agreement, W5/W8 retained-model parity or long-context coverage.",
                       "Sampled process VRAM at context capacity 256, not a transient peak, hardware minimum, multimodal requirement or throughput benchmark."]}


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("bundle", type=Path)
    print(json.dumps(audit(parser.parse_args().bundle), indent=2))

File: research/results/native_decode4_profile_2026_09_13/audit.mjs
// Offline audit of followup4-profile. Does not load binaries or execute a model.
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import cp from 'node:child_process';
import assert from 'node:assert/strict';
const root = path.resolve(process.argv[2]);
const read = p => JSON.parse(fs.readFileSync(path.join(root, p), 'utf8'));
const hash = data => crypto.createHash('sha256').update(data).digest('hex');
const filehash = p => hash(fs.readFileSync(path.join(root, p)));
const location = (name, file = 'report.json') => `stages/${name}/attempt-001/${file}`;
const stage = (name, file) => read(location(name, file));
const sum = values => values.reduce((a, b) => a + b, 0);
const near = (a, b) => assert.ok(Number.isFinite(a) && Number.isFinite(b) &&
  Math.abs(a - b) <= 1e-8 * Math.max(1, Math.abs(b)), `${a} != ${b}`);
const names = ['rotation', 'matrix', 'vocabulary_head', 'embedding'];
const workflow = read('workflow.json'), summary = read('summary.json');
const comparison = read('comparison-b5_v6_s0.json');
const revision = workflow.controls.revision;
assert.equal(revision, '53b8dac6cecf316c22070b2931fee800155072f8');
assert.equal(workflow.controls.dirty_local_test, false);
assert.equal(workflow.controls.diagnostic_profiles, true);
assert.deepEqual(workflow.controls.pilot_contexts, [128]);
assert.deepEqual(workflow.controls.conventional_baselines, []);
assert.equal(workflow.status, 'passed');
assert.equal(summary.status, 'passed');
assert.equal(summary.error, null);
assert.equal(summary.stages.length, 20);
assert.equal(workflow.attempts.length, 20);
assert.equal(new Set(summary.stages.map(x => x.name)).size, 20);
for (const [i, s] of summary.stages.entries()) {
  assert.equal(s.status, 'passed');
  assert.equal(s.name, workflow.attempts[i].name);
  assert.equal(workflow.attempts[i].status, 'passed');
  near(s.elapsed_seconds, workflow.attempts[i].elapsed_seconds);
}
near(sum(summary.stages.map(x => x.elapsed_seconds)), workflow.active_seconds);
near(workflow.active_seconds / 60, summary.active_minutes);
const remoteRoot = workflow.attempts[0].directory.split('/stages/')[0];
let included = 0;
const absent = new Set();
for (const a of workflow.attempts) for (const [name, expected] of Object.entries(a.artifacts || {})) {
  if (!name.startsWith(remoteRoot + '/')) { absent.add(name); continue; }
  const relative = name.slice(remoteRoot.length + 1);
  assert.ok(path.resolve(root, relative).startsWith(root + path.sep));
  assert.equal(filehash(relative), expected.sha256, relative);
  assert.equal(fs.statSync(path.join(root, relative)).size, expected.bytes);
  included++;
}
const runtime = stage('binding-load', 'load.json').runtime_files;
const build = stage('native-build-and-load', 'build-receipt.json');
assert.equal(stage('native-build-and-load', 'cache.json').cache_hit, true);
assert.equal(runtime['librotquant_ggml_test.so'], build.library_sha256);
const code = {...workflow.controls.workflow_sources, ...build.external_sources};
for (const [name, sha] of Object.entries(code)) {
  assert.equal(hash(cp.execFileSync('git', ['show', `${revision}:${name}`])), sha, name);
}
assert.equal(hash(cp.execFileSync('git', ['show', `${revision}:integrations/llama.cpp/rotquant-native-v2.patch`])), build.patch_sha256);
const candidate = stage('decode4-operators-CUDA0');
assert.equal(candidate.decode_cases.length, 54);
for (const r of candidate.decode_cases) assert.equal(r.exact_reference, true);
for (const name of ['candidate-dispatch-preflight', 'operators-CPU', 'operators-CUDA0',
  'decode4-operators-CUDA0', 'whole-model-w6', 'whole-model-w8',
  'decode4-whole-model-w6', 'decode4-whole-model-w8', 'hf-conversion']) {
  const r = stage(name);
  assert.equal(r.passed, true);
  assert.deepEqual(r.runtime_files, runtime);
  for (const c of [...(r.cases || []), ...(r.decode_cases || [])]) assert.equal(c.passed, true);
  if (r.guards) assert.ok(Object.values(r.guards).every(x => x === true));
}
const retained = stage('retained-b5_v6_s0', 'retained/report.json');
const candidateRetained = stage('decode4-retained-b5_v6_s0', 'retained/report.json');
for (const r of [retained, candidateRetained]) {
  assert.equal(r.passed, true);
  assert.equal(r.cpu_fallback_forbidden, true);
  assert.deepEqual(r.runtime_files, runtime);
  assert.equal(r.export_sha256, filehash(location('export-b5_v6_s0', 'export.json')));
  for (const key of ['max_abs_error', 'mean_abs_error', 'mean_kl']) {
    assert.ok(Number.isFinite(r.parity[key]) && r.parity[key] >= 0 && r.parity[key] <= r.parity.thresholds[key]);
  }
  assert.equal(r.parity.top1_agreement, 1);
  assert.equal(r.parity.positions, 16);
  assert.equal(r.parity.exact_generation_probes, 4);
  assert.equal(r.parity.generation_probes, 4);
  assert.ok(Object.values(r.parity.guards).every(x => x === true));
}
assert.deepEqual(candidateRetained.parity, retained.parity);
const refPath = location('pilot-b5_v6_s0-ctx128', 'pilot/report.json');
const ref = read(refPath);
const timings = [], profiles = [];
for (const [label, kind, kernel] of [['pilot', 'pilot', 'reference'], ['decode4', 'pilot', 'decode4'],
  ['profile-reference', 'profile', 'reference'], ['profile-decode4', 'profile', 'decode4']]) {
  const relative = location(`${label}-b5_v6_s0-ctx128`, `${kind}/report.json`);
  const r = read(relative), diagnostic = kind === 'profile';
  assert.equal(r.passed, true);
  assert.equal(r.status, 'passed');
  assert.equal(r.measurement_kind, diagnostic ? 'diagnostic' : 'throughput');
  assert.equal(r.settings.rq3_kernel, kernel);
  assert.equal(r.settings.rq3_profile, diagnostic);
  assert.equal(r.settings.cuda_graphs_disabled, diagnostic);
  assert.equal(r.cpu_fallback_forbidden, true);
  assert.deepEqual(r.runtime_files, runtime);
  assert.deepEqual({...r.settings, rq3_kernel: 'reference', rq3_profile: false, cuda_graphs_disabled: false}, ref.settings);
  for (const key of ['context', 'context_capacity', 'controls', 'prompt_ids_sha256', 'backend',
    'export_sha256', 'probe_sha256', 'complete_model_payload_bytes']) assert.deepEqual(r[key], ref[key]);
  assert.equal(r.parity_report_sha256, filehash(location(
    `${kernel === 'decode4' ? 'decode4-' : ''}retained-b5_v6_s0`, 'retained/report.json')));
  if (label !== 'pilot') assert.equal(r.replay_report_sha256, filehash(refPath));
  assert.equal(r.rows.length, 4);
  for (const [i, row] of r.rows.entries()) {
    assert.equal(row.completed, true);
    assert.equal(row.warmup, i === 0);
    assert.equal(row.input_tokens, 128);
    assert.equal(row.decode_steps, 32);
    assert.equal(row.decode_step_seconds.length, 32);
    assert.deepEqual(row.decode_input_ids, ref.rows[1].decode_input_ids);
    assert.equal(row.decode_input_ids.length, 32);
    assert.ok(row.prefill_seconds > 0 && row.decode_step_seconds.every(x => x > 0));
    near(row.decode_seconds, sum(row.decode_step_seconds));
    near(row.prefill_tokens_per_second, 128 / row.prefill_seconds);
    near(row.decode_tokens_per_second, 32 / row.decode_seconds);
  }
  const measured = r.rows.slice(1);
  assert.equal(r.summary.measured_repetitions, 3);
  near(r.summary.prefill_tokens_per_second, 384 / sum(measured.map(x => x.prefill_seconds)));
  near(r.summary.decode_tokens_per_second, 96 / sum(measured.map(x => x.decode_seconds)));
  if (!diagnostic) {
    timings.push({kernel, ...r.summary, sampled_peak_mib: r.memory.sampled_peak_process_vram_mib});
    continue;
  }
  const receipt = comparison.profiles.find(x => x.kernel === kernel);
  assert.equal(receipt.context, 128);
  assert.equal(receipt.report_sha256, filehash(relative));
  assert.equal(r.native_diagnostics.instrumented, true);
  const binding = r.native_diagnostics.binding;
  assert.equal(binding.lookup, 'execution-library dependency handle');
  assert.equal(path.basename(binding.provider), 'libggml-cuda.so.0');
  assert.equal(binding.provider_sha256, runtime['libggml-cuda.so.0']);
  for (const row of r.rows) for (const phase of ['prefill', 'decode']) {
    const p = row[`${phase}_profile`];
    assert.equal(p.instrumented, true);
    const multiplier = phase === 'decode' ? 32 : 1;
    for (const name of names) {
      assert.ok(Number.isFinite(p.operators[name].milliseconds) && p.operators[name].milliseconds > 0);
      assert.equal(p.operators[name].host_dispatches,
        multiplier * (['matrix', 'rotation'].includes(name) ? 200 : 1));
    }
    assert.equal(p.decode_host_dispatches, kernel === 'decode4' && phase === 'decode' ? 6400 : 0);
    assert.equal(p.tiled_host_dispatches, kernel === 'decode4' && phase === 'prefill' ? 200 : 0);
  }
  for (const name of names) {
    near(r.native_diagnostics.operators[name].milliseconds,
      sum(r.rows.flatMap(x => ['prefill', 'decode'].map(p => x[`${p}_profile`].operators[name].milliseconds))));
  }
  for (const phase of ['prefill', 'decode']) {
    const means = Object.fromEntries(names.map(n => [n, sum(measured.map(x => x[`${phase}_profile`].operators[n].milliseconds)) / 3]));
    const total = sum(Object.values(means));
    for (const n of names) profiles.push({kernel, phase, operator: n,
      mean_gpu_ms_per_repetition: means[n], share_of_custom_event_percent: 100 * means[n] / total,
      ...(phase === 'decode' ? {mean_gpu_ms_per_decode_step: means[n] / 32} : {})});
  }
}
assert.equal(comparison.completed, true);
assert.equal(comparison.comparisons.length, 1);
assert.equal(comparison.profiles.length, 2);
const c = comparison.comparisons[0];
assert.equal(c.reference_report_sha256, filehash(refPath));
assert.equal(c.candidate_report_sha256, filehash(location('decode4-b5_v6_s0-ctx128', 'pilot/report.json')));
near(c.prefill_rate_ratio_to_reference, timings[1].prefill_tokens_per_second / timings[0].prefill_tokens_per_second);
near(c.decode_rate_ratio_to_reference, timings[1].decode_tokens_per_second / timings[0].decode_tokens_per_second);
near(c.sampled_vram_ratio_to_reference, timings[1].sampled_peak_mib / timings[0].sampled_peak_mib);
const result = JSON.stringify({revision, passed_stages: 20, active_minutes: summary.active_minutes,
  checks: {included_artifact_references: included, external_artifacts_absent: absent.size, producer_files: Object.keys(code).length},
  timings, comparisons: comparison.comparisons, profiles, retained_parity: retained.parity,
  boundary: 'Offline hash/arithmetic and reported-gate audit; no GPU replay. Profiles are custom-op CUDA event time with graphs disabled, not whole-model wall shares, bandwidth or task accuracy.'}, null, 2) + '\n';
if (process.argv[3]) fs.writeFileSync(process.argv[3], result, {flag: 'wx'});
else process.stdout.write(result);
